[iOS] Third-party video apps lack standardized caption customization for FCC compliance

Phinny01 opened the pull request:
#### f27610320c33bed78f1a2a19e2e8a3652c039e66
<pre>
[iOS] Third-party video apps lack standardized caption customization for FCC compliance
<a href="https://bugs.webkit.org/show_bug.cgi?id=302359">https://bugs.webkit.org/show_bug.cgi?id=302359</a>
<a href="https://rdar.apple.com/164670301">rdar://164670301</a>

Reviewed by Jer Noble.

Replace existing media controls captions styles menu for iOS with the new AVKit SPI to meet FCC
compliance. Add SPI declaration to AVKitSPI.h and soft-link it in WKCaptionStyleMenuControllerIOS

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h:
* Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuControllerInternal.h: Copied from Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h.
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant showMediaControlsContextMenu:items:frameInfo:identifier:completionHandler:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.h: Copied from Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h.
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm: Added.
(-[_WKCaptionStyleMenuControllerAVKit init]):
(-[_WKCaptionStyleMenuControllerAVKit rebuildMenu]):
(-[_WKCaptionStyleMenuControllerAVKit isAncestorOf:]):
(-[_WKCaptionStyleMenuControllerAVKit legibleMenuController:didRequestCaptionPreviewForProfileID:]):
(-[_WKCaptionStyleMenuControllerAVKit findAndDismissContextMenus]):
(-[_WKCaptionStyleMenuControllerAVKit searchForContextMenuInteractionsInView:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
(+[WKCaptionStyleMenuController menuController]):
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController menu]):
(-[WKCaptionStyleMenuController setMenu:]):
(-[WKCaptionStyleMenuController interaction]):
(-[WKCaptionStyleMenuController setInteraction:]):
(-[WKCaptionStyleMenuController profileActionSelected:]):
(-[WKCaptionStyleMenuController notifyMenuWillOpen]):
(-[WKCaptionStyleMenuController notifyMenuDidClose]):
(-[WKCaptionStyleMenuController contextMenuInteraction:configurationForMenuAtLocation:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(-[TestContextMenuDelegate contextMenuInteraction:configurationForMenuAtLocation:]):
(-[TestAVKitDelegate legibleMenuController:didRequestCaptionPreviewForProfileID:]):
(TestWebKitAPI::CaptionPreferenceTests::ensureController):
(TestWebKitAPI::CaptionPreferenceTests::runAndWaitForCaptionPreviewRequested):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, AVKitMenuControllerIntegration)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ProfileIDSavingAndRestoration)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ContextMenuDismissalSearch)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuRebuildingWithAVKit)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuAncestryCheck)):

Canonical link: <a href="https://commits.webkit.org/303293@main">https://commits.webkit.org/303293@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fa46780023630ecb52ccbe4f05046ea290ecb27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42939 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139440 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e12e9cfd-b62a-44d5-b729-64245cce6519) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133798 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4360 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4182 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100842 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134874 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118152 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81632 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8ae5f77e-5709-45bd-96d4-2a7c1c7a87dc) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82660 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123991 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36275 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142085 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130435 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/4090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36855 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109213 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4171 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109382 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27707 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3100 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114432 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57311 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4143 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32830 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163402 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3975 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67590 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42481 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4103 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/Configurations/AllowedSPI.toml:
```diff
@@ -227,3 +227,13 @@ selectors = [
+
+[[temporary-usage]]
+request = "rdar://OOPS!"
+cleanup = "rdar://OOPS!"
```
Please turn the OOPS-es into actual API request/cleanup bugs before you land. I thought we had a style check for this, but perhaps it's not working

Phinny01 replied:
Done

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
I think we should probably `return nil;` here.

Phinny01 replied:
Done

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
You can skip this `if()` check.

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
These two are the same as the parent class's ivars, so they can just be dropped.

Phinny01 replied:
You're right. I added them to the header file instead

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
If you remove the `if()` check here, the call to `-buildMenuOfType:` will return `nil` and set `_menu` to `nil`, which is probably what we want to do here.

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
This isn't _quite_ right, but it's close enough for now. We'd need a way to walk up the UIMenu hierarchy, and that doesn't really exist.

Phinny01 replied:
Do we need UIKit to make an API to be able to walk up the UIMenu hierarchy some time in the future?

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm:
These all exist (and are the same) in the parent class, so they can be dropped.

Phinny01 replied:
right

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
I don't think this is the right style for `@interface` blocks, but I could be wrong.

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
This looks wrong somehow. It probably shouldn't be necessary to override `+alloc`, and the way it's working here means the object was already pre-`-init`-ed inside of `+alloc`.

jernoble commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
Unnecessary whitespace.

jernoble commented on Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
```diff
@@ -62,8 +62,11 @@ NS_ASSUME_NONNULL_END
-#if PLATFORM(IOS_FAMILY)
+#if HAVE(AVLEGIBLEMEDIAOPTIONSMENUCONTROLLER)
+#import <AVKit/AVLegibleMediaOptionsMenuController.h>
+#endif
```
Lets replace this with a `#if __has_include`.

jernoble commented on Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h:
A better way to do this would be to add a new `_WKCaptionStyleMenuControllerInternal.h` that has the class extension and the instance variables.

Phinny01 replied:
I moved the ivars into _WKCaptionStyleMenuControllerInternal.h with a class extension, but the export checker is now complaining:
no declaration found for exported symbol '(ObjC IVar) WKCaptionStyleMenuController._menu'.
I’ve added the internal header as a Private header on the Webkit target. Is there another header list / script I should register it with for the symbol checker?

jernoble replied:
Good question; I haven't dealt with this particular error message before.

jernoble reviewed (approved):
r=me with nits

Commit:
[iOS] Third-party video apps lack standardized caption customization for FCC compliance
https://bugs.webkit.org/show_bug.cgi?id=302359
rdar://164670301

Reviewed by Jer Noble.

Replace existing media controls captions styles menu for iOS with the new AVKit SPI to meet FCC
compliance. Add SPI declaration to AVKitSPI.h and soft-link it in WKCaptionStyleMenuControllerIOS

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebKit/Configurations/AllowedSPI.toml:
* Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h:
* Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuControllerInternal.h: Copied from Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h.
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant showMediaControlsContextMenu:items:frameInfo:identifier:completionHandler:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.h: Copied from Source/WebKit/UIProcess/Cocoa/_WKCaptionStyleMenuController.h.
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerAVKit.mm: Added.
(-[_WKCaptionStyleMenuControllerAVKit init]):
(-[_WKCaptionStyleMenuControllerAVKit rebuildMenu]):
(-[_WKCaptionStyleMenuControllerAVKit isAncestorOf:]):
(-[_WKCaptionStyleMenuControllerAVKit legibleMenuController:didRequestCaptionPreviewForProfileID:]):
(-[_WKCaptionStyleMenuControllerAVKit findAndDismissContextMenus]):
(-[_WKCaptionStyleMenuControllerAVKit searchForContextMenuInteractionsInView:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
(+[WKCaptionStyleMenuController menuController]):
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController menu]):
(-[WKCaptionStyleMenuController setMenu:]):
(-[WKCaptionStyleMenuController interaction]):
(-[WKCaptionStyleMenuController setInteraction:]):
(-[WKCaptionStyleMenuController profileActionSelected:]):
(-[WKCaptionStyleMenuController notifyMenuWillOpen]):
(-[WKCaptionStyleMenuController notifyMenuDidClose]):
(-[WKCaptionStyleMenuController contextMenuInteraction:configurationForMenuAtLocation:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(-[TestContextMenuDelegate contextMenuInteraction:configurationForMenuAtLocation:]):
(-[TestAVKitDelegate legibleMenuController:didRequestCaptionPreviewForProfileID:]):
(TestWebKitAPI::CaptionPreferenceTests::ensureController):
(TestWebKitAPI::CaptionPreferenceTests::runAndWaitForCaptionPreviewRequested):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, AVKitMenuControllerIntegration)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ProfileIDSavingAndRestoration)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ContextMenuDismissalSearch)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuRebuildingWithAVKit)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, MenuAncestryCheck)):

Canonical link: https://commits.webkit.org/303293@main
